lib/functions/transfer.js

```diff
@@ -82,9 +82,31 @@ class Transfer extends AsyncGeneratorFunction {
                         headers: Object.assign({
                             [HTTP.HEADER.CONTENT_LENGTH]: blob.size,
                             [HTTP.HEADER.CONTENT_TYPE]: transferPart.metadata.contentType
-                        }, transferPart.headers)
+                        }, transferPart.targetHeaders)
                     });
                 }, this.options);
+            } else if (targetUrl && isFileProtocol(targetUrl) && transferPart.source.url) {
+                await retry(async () => {
+                    const totalSize = transferPart.metadata.contentLength;
+                    const response = await streamGet(transferPart.source.url, {
+                        headers: Object.assign({
+                            [HTTP.HEADER.RANGE]: `${HTTP.RANGE.BYTES}=${contentRange.low}-${contentRange.high}`
+                        }, transferPart.sourceHeaders)
+                    });
+                    const contentLength = response.headers.get(HTTP.HEADER.CONTENT_LENGTH);
+
+                    // there have been cases where the server does not honor the range header.
+                    // since we'll be reading the entire response into memory, protect against accidentally
+                    // reading a very large amount of data into memory. For example, if the server responds with
+                    // an entire 1GB file even though we only requested a portion of that file, avoid reading
+                    // the whole 1GB into memory.
+                    const actualContentDifference = contentLength - contentRange.length;
+                    if (actualContentDifference > 0) {
+                        throw Error(`Server does not seem to have respected Range header. Actual content length is ${contentLength}B, which is ${actualContentDifference}B larger than requested range of ${contentRange.low}-${contentRange.high}.`);
+                    }
+
+                    await this.randomFileAccess.writeStream(targetUrl, contentRange, response.body, totalSize);
```

Comment:
`Transfer` should be dealing with reading it in to blocks. It's purpose is map the streams in to block I/O.